api/middleware/logger.go

```diff
@@ -24,10 +26,35 @@ type LoggingResponseWriter struct {
 	errBody []byte // the body in case it is an error
 }
 
+func decompressGzip(b []byte) ([]byte, error) {
+	reader, err := gzip.NewReader(bytes.NewReader(b))
+	if err != nil {
+		return nil, err
+	}
+
+	return ioutil.ReadAll(reader)
+}
+
 func (rw *LoggingResponseWriter) Write(b []byte) (int, error) {
 	if rw.ResponseWriter.Status() >= 400 {
-		rw.errBody = make([]byte, len(b))
-		copy(rw.errBody, b)
+		if rw.Header().Get("Content-Encoding") == "gzip" {
+			d, err := decompressGzip(b)
+			if err != nil {
+				log.Errorf("Decompressing gzip body failed: %s", err.Error())
+				// In some cases, decompression fails because the body is in fact
+				// not compressed, save the body as is.
+				// This happens when the gziper middleware is enabled because it
+				// sets the 'Content-Encoding' header to 'gzip' before it actually
+				// compresses the body.
+				rw.errBody = make([]byte, len(b))
```

Comment:
`rw.Write(b)` might be called multiple times.  This will result in the errBody only containing the end of the data.  
If response resulted in an Error and the payload is really large, we probably dont want all of it, but we will want the first part of it.

Reply:
Fixed with ff4b0e5 by buffering up all the entire response. For now I think we can store it all. Having it all might be necessary in the `gzip` case.